### PR TITLE
Pamd Module

### PR DESCRIPTION
**Pull request by kevensen**
##### ISSUE TYPE

- New Module Pull Request
##### COMPONENT NAME

pamd
##### ANSIBLE VERSION

```
ansible 2.1.0.0
```
##### SUMMARY

While lineinfile and replace are very useful, I felt that a module specific to updating pam.d services was necessary.

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @kevensen for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by alikins** [comment]
misc flake8 stuff:

```
pamd.py:152:0: W293 blank line contains whitespace
pamd.py:209:0: W293 blank line contains whitespace
pamd.py:287:0: W293 blank line contains whitespace
pamd.py:303:0: W293 blank line contains whitespace
pamd.py:134:0: E265 block comment should start with '# '
pamd.py:188:0: E265 block comment should start with '# '
pamd.py:135:0: E302 expected 2 blank lines, found 1
pamd.py:189:0: E302 expected 2 blank lines, found 1
pamd.py:213:0: E302 expected 2 blank lines, found 1
pamd.py:333:4: E301 expected 1 blank line, found 0
pamd.py:353:16: E221 multiple spaces before operator
pamd.py:354:16: E221 multiple spaces before operator
pamd.py:362:16: E221 multiple spaces before operator
pamd.py:353:16: E251 unexpected spaces around keyword / parameter equals
pamd.py:354:16: E251 unexpected spaces around keyword / parameter equals
pamd.py:362:16: E251 unexpected spaces around keyword / parameter equals
pamd.py:361:17: E221 multiple spaces before operator
pamd.py:361:17: E251 unexpected spaces around keyword / parameter equals
pamd.py:355:19: E221 multiple spaces before operator
pamd.py:355:19: E251 unexpected spaces around keyword / parameter equals
pamd.py:357:20: E221 multiple spaces before operator
pamd.py:357:20: E251 unexpected spaces around keyword / parameter equals
pamd.py:352:21: E251 unexpected spaces around keyword / parameter equals
pamd.py:356:23: E221 multiple spaces before operator
pamd.py:358:23: E221 multiple spaces before operator
pamd.py:352:23: E251 unexpected spaces around keyword / parameter equals
pamd.py:356:23: E251 unexpected spaces around keyword / parameter equals
pamd.py:358:23: E251 unexpected spaces around keyword / parameter equals
pamd.py:359:27: E221 multiple spaces before operator
pamd.py:359:27: E251 unexpected spaces around keyword / parameter equals
pamd.py:364:27: E251 unexpected spaces around keyword / parameter equals
pamd.py:360:28: E251 unexpected spaces around keyword / parameter equals
pamd.py:364:29: E251 unexpected spaces around keyword / parameter equals
pamd.py:353:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:354:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:355:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:356:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:357:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:358:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:359:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:360:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:361:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:362:30: E251 unexpected spaces around keyword / parameter equals
pamd.py:317:36: E201 whitespace after '['
pamd.py:354:60: E201 whitespace after '['
pamd.py:357:61: E201 whitespace after '['
pamd.py:361:80: E201 whitespace after '['
pamd.py:317:115: E202 whitespace before ']'
pamd.py:122:130: E501 line too long (135 > 130 characters)
pamd.py:361:130: E501 line too long (143 > 130 characters)
pamd.py:108:130: E501 line too long (145 > 130 characters)
pamd.py:118:130: E501 line too long (153 > 130 characters)
pamd.py:120:130: E501 line too long (154 > 130 characters)
pamd.py:223:130: E501 line too long (154 > 130 characters)
pamd.py:253:130: E501 line too long (154 > 130 characters)
pamd.py:275:130: E501 line too long (154 > 130 characters)
pamd.py:294:130: E501 line too long (154 > 130 characters)
pamd.py:313:130: E501 line too long (154 > 130 characters)
pamd.py:393:130: E501 line too long (158 > 130 characters)
pamd.py:397:130: E501 line too long (158 > 130 characters)
pamd.py:112:130: E501 line too long (160 > 130 characters)
pamd.py:110:130: E501 line too long (161 > 130 characters)
pamd.py:116:130: E501 line too long (171 > 130 characters)
pamd.py:276:130: E501 line too long (210 > 130 characters)
pamd.py:257:130: E501 line too long (212 > 130 characters)
```

File: system/pamd.py [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+            
+            try:
+                if self.rule_module_args is not None and type(self.rule_module_args) is list:
+                    self.rule_module_args = rule_module_args
+                elif self.rule_module_args is not None and type(self.rule_module_args) is str:
+                    self.rule_module_args = rule_module_args.split(' ')
+            except AttributeError:
+                self.rule_module_args = []
+
+    def __parsline__(self):
```
Comment (alikins) [inline system/pamd.py]: That is an unusual method name.

Why the leading '__' ? Why the trailing '__'?

File: system/pamd.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+        self.path = path
+        self.name = name
+        self.check = ansible.check_mode
+        self.ansible = ansible
+        self.fname = self.path + "/" + self.name
+        self.preamble = []
+        self.rules = []
+
+        if os.path.exists(self.fname):
+            for line in fileinput.FileInput(self.fname, inplace=False):
```
Comment (alikins) [inline system/pamd.py]: I wouldn't use fileinput here.

This module should never need to read from stdin directly, and sys.argv is handled by the module base class so
fileinput seems unnecessary.

As far as I can tell, a regular file object via open() should be sufficient here.

File: system/pamd.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+#The PamdRule class encapsulates a rule in a pam.d service
+class PamdRule(object):
+    def __init__(self, ansible, stringline=None, rule_type=None,
+                 rule_control=None, rule_module_path=None,
+                 rule_module_args=None):
+
+        if stringline is not None and (rule_type is not None or
+                                       rule_control is not None or
+                                       rule_module_path is not None):
+            ansible.fail_json(msg='This is bizarre.  You can only pass' +
```
Comment (alikins) [inline system/pamd.py]: Maybe this should be two separate classes?  Or a single model class and methods for constructing it from either type/control/path or a stringline?

File: system/pamd.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+        else:
+            self.rule_type = rule_type
+            self.rule_control = rule_control
+            self.rule_module_path = rule_module_path
+            
+            try:
+                if self.rule_module_args is not None and type(self.rule_module_args) is list:
+                    self.rule_module_args = rule_module_args
+                elif self.rule_module_args is not None and type(self.rule_module_args) is str:
+                    self.rule_module_args = rule_module_args.split(' ')
```
Comment (alikins) [inline system/pamd.py]: the split(' ') makes trailing and repeated spaces significant. Is that intentional (ie, do they need to be preserved to construct the pam line correctly?)

I'd lean towards just .split() if not.

``` python
>>>b = 'fg dfg dfg dfg  sd  '
>>> b.split(' ')
['fg', 'dfg', 'dfg', 'dfg', '', 'sd', '', '']
>>> b.split()
['fg', 'dfg', 'dfg', 'dfg', 'sd']
```

File: system/pamd.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+                elif not line.startswith('#') and not line.isspace():
+                    self.rules.append(PamdRule(ansible, stringline=line.rstrip()))
+            fileinput.close()
+        else:
+            self.ansible.fail_json(msg='Unable to locate PAM module file %s' % self.fname)
+    
+    def __str__(self):
+        return self.fname
+
+class PamdUtil(object):
```
Comment (alikins) [inline system/pamd.py]: Since this is all static methods, making them module scope methods would be more traditional.
But the class is fine as well.

**Comment by kevensen** [comment]
My Python is really rusty so I appreciate the feedback.  I'll get this cleaned up and committed.

**Comment by kevensen** [comment]
Hi @alikins.  When you get a chance, can you give this module another look?  Please let me  know if there's anything else I need to do to have this pull request accepted.

**Pull request closed without merging**